### PR TITLE
solana/program_v2: Add tests to gatekeeper.rs

### PR DESCRIPTION
**Pull request by JulianSpring**
Added tests to the update functions used in update_gatekeeper IDCOM-2316

### CODE REVIEW

File: solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs [hunk 1]
```diff
@@ -313,3 +299,421 @@ bitflags! {
...
+            true,
+            true,
+            lamports,
+            data,
+            owner,
+            false,
+            Epoch::default(),
+        );
+
+        let authority = Signer::try_from(&account_info).unwrap();
```
Comment (JulianSpring) [inline solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs:327]: Struggling to make this int a helper function due to ownership issues. Would love some help here

Comment (JulianSpring) [reply solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs]: @tbarri helped with a great little function!

**Review by william-brooks — APPROVED** [review]

File: solana/program_v2/programs/gateway_v2/src/instructions/network/update_gatekeeper.rs [hunk 2]
```diff
@@ -12,15 +13,29 @@ pub fn update_gatekeeper(
...
+impl UpdateGatekeeperData {
+    fn can_update_auth_keys(&self, gatekeeper: &Gatekeeper, authority: &Signer) -> bool {
+        (self.auth_keys.remove.is_empty() && self.auth_keys.add.is_empty())
+            || gatekeeper.can_access(authority, GatekeeperKeyFlags::AUTH)
+    }
+
+    fn can_update_fees(&self, gatekeeper: &Gatekeeper, authority: &Signer) -> bool {
+        (self.token_fees.add.is_empty() && self.token_fees.remove.is_empty())
+            || gatekeeper.can_access(authority, GatekeeperKeyFlags::ADJUST_FEES)
+    }
```
[the remark this answers is not in the document]
Comment (JulianSpring) [reply solana/program_v2/programs/gateway_v2/src/instructions/network/update_gatekeeper.rs:32]: Ah thanks you are right, ill add in Adjust fees. Think we need to have a integration test to test the constraints? 😅

File: solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs [hunk 3]
```diff
@@ -313,3 +299,306 @@ bitflags! {
...
+
+            // Act
+            gatekeeper
+                .add_and_remove_fees(&update_gatekeeper_fees)
+                .unwrap();
+
+            // Assert
+            assert_eq!(gatekeeper.token_fees[0], expected_fees);
+        });
+    }
```
[the remark this answers is not in the document]
Comment (JulianSpring) [reply solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs:491]: Yeah you are right!

File: solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs [hunk 4]
```diff
@@ -313,3 +299,306 @@ bitflags! {
...
+
+            // Act
+            gatekeeper
+                .add_and_remove_auth_keys(&update_gatekeeper_fees, &authority)
+                .unwrap();
+
+            // Assert
+            assert_eq!(gatekeeper.auth_keys[0], original_auth_key);
+            assert_eq!(gatekeeper.auth_keys[1], expected_gatekeeper);
+        })
```
[the remark this answers is not in the document]
Comment (JulianSpring) [reply solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs:390]: Yep! Thanks!

File: solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs [hunk 5]
```diff
@@ -313,3 +299,306 @@ bitflags! {
...
+    use crate::instructions::network::UpdateGatekeeperFees;
+    use solana_program::clock::Epoch;
+
+    #[test]
+    fn test_can_access_with_auth_authority() {
+        with_signer(|authority| {
+            let gatekeeper = make_gatekeeper(
+                &Some(authority.clone()),
+                &Some(authority.key()),
+                &GatekeeperKeyFlags::AUTH,
```
[the remark this answers is not in the document]
Comment (JulianSpring) [reply solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs:316]: Yeah that is probably better, will take a look

File: solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs [hunk 6]
```diff
@@ -313,3 +299,306 @@ bitflags! {
 impl OnChainSize for GatekeeperKeyFlags {
     const ON_CHAIN_SIZE: usize = OC_SIZE_U32;
 }
+
+#[cfg(test)]
+mod gatekeeper_tests {
```
[the remark this answers is not in the document]
Comment (JulianSpring) [reply solana/program_v2/programs/gateway_v2/src/state/gatekeeper.rs:304]: Thanks!

**Pull request merged**